pkg/cache/cache.go

```diff
@@ -0,0 +1,251 @@
+//   - [NewTransformer]: A cache that transforms data from one format to
+//     another. It's only refreshed when the source changes.
+//   - [NewMerger]: A cache that aggregates multiple caches into one.
+//     It's only refreshed when the source changes.
+//   - [Replaceable]: A cache adapter that can ben atomically
+//     replaced with a new one, and saves the previous results in case an
+//     error pops-up.
+//
+// # Atomicity
+//
+// Most of the operations are not atomic/thread-safe, except for
+// [Replaceable.Replace] which can be performed while the objects
+// are being read.
+package cache
+
+import (
+	"fmt"
+	"sync/atomic"
+)
+
+// Result is the data returned from a call to a cache. It can either be
+// created with [NewResultData] if the call was a success, or
+// [NewResultError] if the call resulted in an error.
+type Result[T any] struct {
+	data T
+	etag string
+	err  error
+}
+
+// NewResultData creates a new [Result] for a successful operation.
+func NewResultData[T any](data T, etag string) Result[T] {
+	return Result[T]{
+		data: data,
+		etag: etag,
+	}
+}
+
+// NewResultError creates a new [Result] when an error has happened.
+func NewResultError[T any](err error) Result[T] {
+	return Result[T]{
+		err: err,
+	}
+}
+
+// Result can be treated as a [Interface] if necessary.
+func (r Result[T]) Data() Result[T] {
+	return r
+}
+
+// Split the result into its data, etag and error components.
+func (r Result[T]) Split() (T, string, error) {
+	return r.data, r.etag, r.err
+}
+
+// Interface is a cache that performs an action that will be cached. It
+// also returns an "etag" identifier to version the cache, so that the
+// caller can know if they have the most recent version of the cache
+// (and can decide to cache some operation based on that).
+//
+// The [NewMerger] and [NewTransformer] automatically handle
+// that for you by checking if the etag is updated before calling the
+// merging or transforming function.
+type Interface[T any] interface {
+	// Returns the cached data, as well as an "etag" to identify the
+	// version of the cache, or an error if something happened.
+	Data() Result[T]
+}
+
+// T is the source type, V is the destination type.
+type merger[K comparable, T, V any] struct {
+	mergeFn      func(map[K]Result[T]) Result[V]
+	caches       map[K]Interface[T]
+	cacheResults map[K]Result[T]
+	result       Result[V]
+}
+
+// NewMerger creates a new merge cache, a cache that merges the result
+// of other caches. The function only gets called if any of the
+// dependency has changed.
+//
+// If any of the dependency returned an error before, or any of the
+// dependency returned an error this time, or if the mergeFn failed
+// before, then the function is reran.
+//
+// The caches and results are mapped by string so that associated data
+// can be retrieved.
+func NewMerger[K comparable, T, V any](mergeFn func(results map[K]Result[T]) Result[V], caches map[K]Interface[T]) Interface[V] {
```

Comment:
Realistically what kind of types will we have for `K comparable?`

Reply:
I only have strings so far, but it seemed odd as I was adding the generics everywhere, not to allow a different type of key? I can revert that back to string if you want.

Reply:
This is fine, I was just curious about the reasoning for the comparable, and maintaining the generic interface is a good reason.